lib/SILOptimizer/Mandatory/SILGenCleanup.cpp

```diff
@@ -104,39 +108,219 @@ struct SILGenCleanup : SILModuleTransform {
+  for (auto &block : *function) {
+    if (&block == function->getEntryBlock()) {
+      // TODO: [fix_silgen_destroy_unreachable] Remove this condition.
+      continue;
+    }
+    if (!block.pred_empty())
+      continue;
+    if (!backward.isVisited(&block))
+      continue;
+    roots.push_back(&block);
+  }
+
+  // Complex case: unreachable loops.
+  //
+  // Iteratively (the first time, these are the roots discovered in "Simple
+  // case" above), determine which blocks are forward-reachable from roots.
+  // Then, look for a block that is backwards-reachable from dead-end blocks
+  // but not forwards-reachable from roots so far discovered.  If one is found,
+  // it is forwards-reachable from an unreachable loop.  Walk backwards from
+  // that block to find a representative block in the loop.  Add that
+  // representative block to roots and iterate.  If no such block is found, all
+  // roots have been found.
+  BasicBlockWorklist forward(function);
+  for (auto *root : roots) {
+    forward.push(root);
+  }
+  bool changed = false;
+  auto iterator = function->begin();
+  do {
+    changed = false;
+
+    // Propagate forward-reachability from roots discovered so far.
+    while (auto *block = forward.pop()) {
+      for (auto *successor : block->getSuccessorBlocks()) {
+        forward.pushIfNotVisited(successor);
+      }
+    }
+    // Any block in `backward` but not in `forward` is forward-reachable from an
+    // unreachable loop.
+    if (auto *target =
+            findFirstBlock(iterator, /*include=*/backward.getVisited(),
+                           /*exclude=*/forward.getVisited())) {
+      // Find the first unreachable loop it's forwards-reachable from.
+      auto *loop = findFirstLoop(function, target);
+      ASSERT(loop);
+      forward.push(loop);
+      roots.push_back(loop);
+      changed = true;
+    }
+  } while (changed);
+}
+
 bool SILGenCleanup::completeOSSALifetimes(SILFunction *function) {
   if (!getModule()->getOptions().OSSACompleteLifetimes)
     return false;
 
+  LLVM_DEBUG(llvm::dbgs() << "Completing lifetimes in " << function->getName()
+                          << "\n");
+
+  BasicBlockWorklist deadends(function);
+  DeadEndBlocks *deba = getAnalysis<DeadEndBlocksAnalysis>()->get(function);
+  for (auto &block : *function) {
+    if (deba->isDeadEnd(&block))
+      deadends.push(&block);
+  }
+
+  if (deadends.empty()) {
+    // There are no dead-end blocks, so there are no lifetimes to complete.
+    // (SILGen may emit incomplete lifetimes, but not underconsumed lifetimes.)
+    return false;
+  }
+
+  // Lifetimes must be completed in unreachable blocks that are reachable via
+  // backwards walk from dead-end blocks.  First, check whether there are any
+  // unreachable blocks.
+  ReachableBlocks reachableBlocks(function);
+  reachableBlocks.compute();
+  StackList<SILBasicBlock *> roots(function);
+  if (!reachableBlocks.hasUnreachableBlocks()) {
```

Comment:
It's probably not worth checking for the special case of no unreachable blocks in the function. `collectReachableRoots` will yield the same result with approximately the same effort.

Reply:
Well, you could do the work of loop #2 in loop #1, which reduces the overhead to a single loop, which is IMO negligible.

Reply:
While it might be negligible, it's still worse and it's already written this way.  Why change it to something negligibly worse?

Reply:
The benefit would be to reduce the complexity of the source code (which is always my goal # 1): no need for `hasUnreachableBlocks`, less different branches to be taken, etc.

> it's still worse

I don't think so. We are iterating over blocks and instructions so many times that the overhead is for sure not measurable.

> it's already written this way

This is not a good argument 🙂

All that being said, this is nitpicking. If you prefer to keep it this way, I'm fine with it.